src/track/globaltrackcache.cpp

```diff
@@ -139,41 +154,111 @@ GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
 
 //static
 void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(!s_pInstance);
     s_pInstance = new GlobalTrackCache(pEvictor);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
     s_pInstance = nullptr;
     delete pInstance;
 }
 
 //static
 void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+    DEBUG_ASSERT(pTrack);
+    // Any access to pTrack is forbidden!! Due to race condition
+    // this pointer might already have been deleted! Only the
+    // cache knows this.
+    if (s_pInstance) {
+        s_pInstance->evictAndDelete(pTrack);
+    } else {
+        // Simply delete unreferenced tracks when the cache is no
+        // longer available. This might but should not happen.
+        kLogger.warning()
+                << "Deleting uncached track";
```

Comment:
It this warning correct? Or does it also happen with an original cached track where the cache is gone?

Reply:
I actually have this message two times in my log, playing only on track.

Reply:
This happens only when shutting down Mixxx. When destroying the GlobalTrackCache we cannot ensure that other components have released all their internally cached Track pointers. Please note that the GlobalTrackCache has to be destroyed before the TrackDAO gets destroyed if you don't want to lose any modifications.